SimPEG/EM/NSEM/SurveyNSEM.py

```diff
@@ -0,0 +1,207 @@
+            # Get the type and the value for the DataNSEM object as a list
+            typeList = [[rx.orientation,rx.component,self[src,rx]] for rx in src.rxList]
+            # Insert the values to the temp array
+            for nr,(k,c,val) in enumerate(typeList):
+                key = 'z' + k + c[0]
+                tArrRec[key] = mkvc(val,2)
+            # Masked array
+            # mArrRec = np.ma.MaskedArray(_rec_to_ndarr(tArrRec),mask=np.isnan(_rec_to_ndarr(tArrRec))).view(dtype=tArrRec.dtype)
+
+            try:
+                outTemp = recFunc.stack_arrays((outTemp,tArrRec))
+            except NameError:
+                outTemp = tArrRec.copy()
+
+            if 'RealImag' in returnType:
+                outArr = outTemp.copy()
+            elif 'Complex' in returnType:
+                # Add the real and imaginary to a complex number
+                outArr = np.empty(outTemp.shape,dtype=dtCP)
+                for comp in ['freq','x','y','z']:
+                    outArr[comp] = outTemp[comp].copy()
+                for comp in ['zxx','zxy','zyx','zyy','tzx','tzy']:
+                    outArr[comp] = outTemp[comp+'r'].copy() + 1j*outTemp[comp+'i'].copy()
+            else:
+                raise NotImplementedError('{:s} is not implemented, as to be RealImag or Complex.')
+
+        # Return
+        return outArr
+
+    @classmethod
+    def fromRecArray(cls, recArray, srcType='primary'):
+        """
+        Class method that reads in a numpy record array to NSEMdata object.
+
+        :param numpy.recarray recArray: Record array with the data. Has to have
+            ('freq','x','y','z') columns and some ('zxx','zxy','zyx','zyy','tzx','tzy')
+        :param string srcType: The type of SimPEG.EM.NSEM.SrcNSEM to be used
+
+        """
+        if srcType=='primary':
+            src = polxy_1Dprimary
+        elif srcType=='total':
+            src = polxy_1DhomotD
+        else:
+            raise NotImplementedError('{:s} is not a valid source type for NSEMdata')
+
+        # Find all the frequencies in recArray
+        uniFreq = np.unique(recArray['freq'].copy())
+        srcList = []
+        dataList = []
+        for freq in uniFreq:
+            # Initiate rxList
+            rxList = []
+            # Find that data for freq
+            dFreq = recArray[recArray['freq'] == freq].copy()
+            # Find the impedance rxTypes in the recArray.
+            rxTypes = [ comp for comp in recArray.dtype.names if (len(comp)==4 or len(comp)==3) and 'z' in comp]
+            for rxType in rxTypes:
+                # Find index of not nan values in rxType
+                notNaNind = ~np.isnan(dFreq[rxType].copy())
+                if np.any(notNaNind): # Make sure that there is any data to add.
+                    locs = _rec_to_ndarr(dFreq[['x','y','z']][notNaNind].copy())
+                    if dFreq[rxType].dtype.name in 'complex128':
```

Comment:
This is done to remove the nan's from the recarray and only insert locations that actually have data into the Data class. And if there no data at all, there won't be a receiver added.